File: backend/controllers/commentController.js
const Comment = require('../models/Comment');

const createComment = async (req, res) => {
  try {
    const { content } = req.body;
    const { recipeId } = req.params;

    const comment = await Comment.create({
      content,
      author: req.user._id,
      recipe: recipeId
    });

    await comment.populate('author', 'name');
    res.status(201).json(comment);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const getComments = async (req, res) => {
  try {
    const { recipeId } = req.params;
    
    const comments = await Comment.find({ recipe: recipeId })
      .populate('author', 'name')
      .sort({ createdAt: -1 });

    res.json(comments);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const deleteComment = async (req, res) => {
  try {
    const comment = await Comment.findById(req.params.id);
    
    if (!comment) {
      return res.status(404).json({ message: 'Comment not found' });
    }

    if (comment.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized' });
    }

    await Comment.findByIdAndDelete(req.params.id);
    res.json({ message: 'Comment deleted' });
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

module.exports = { createComment, getComments, deleteComment };

File: springboot-backend/src/main/java/com/foodstagram/controller/CommentController.java
package com.foodstagram.controller;

import com.foodstagram.model.Comment;
import com.foodstagram.model.Recipe;
import com.foodstagram.model.User;
import com.foodstagram.repository.CommentRepository;
import com.foodstagram.repository.RecipeRepository;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import java.util.*;

@RestController
@RequestMapping("/api/comments")
public class CommentController {
    private final CommentRepository commentRepository;
    private final RecipeRepository recipeRepository;

    public CommentController(CommentRepository commentRepository, RecipeRepository recipeRepository) {
        this.commentRepository = commentRepository;
        this.recipeRepository = recipeRepository;
    }

    @GetMapping("/{recipeId}")
    public ResponseEntity<?> getComments(@PathVariable Long recipeId) {
        Recipe recipe = recipeRepository.findById(recipeId).orElse(null);
        if (recipe == null) return ResponseEntity.notFound().build();
        List<Comment> comments = commentRepository.findByRecipeOrderByCreatedAtDesc(recipe);
        return ResponseEntity.ok(comments.stream().map(this::toMap).toList());
    }

    @PostMapping("/{recipeId}")
    public ResponseEntity<?> createComment(@PathVariable Long recipeId,
                                            @RequestBody Map<String, String> body,
                                            @AuthenticationPrincipal User user) {
        Recipe recipe = recipeRepository.findById(recipeId).orElse(null);
        if (recipe == null) return ResponseEntity.notFound().build();

        Comment comment = new Comment();
        comment.setContent(body.get("content"));
        comment.setAuthor(user);
        comment.setRecipe(recipe);
        commentRepository.save(comment);
        return ResponseEntity.status(201).body(toMap(comment));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteComment(@PathVariable Long id, @AuthenticationPrincipal User user) {
        Comment comment = commentRepository.findById(id).orElse(null);
        if (comment == null) return ResponseEntity.notFound().build();
        if (!comment.getAuthor().getId().equals(user.getId()))
            return ResponseEntity.status(403).body(Map.of("message", "Not authorized"));
        commentRepository.delete(comment);
        return ResponseEntity.ok(Map.of("message", "Comment deleted"));
    }

    private Map<String, Object> toMap(Comment c) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("_id", c.getId());
        map.put("id", c.getId());
        map.put("content", c.getContent());
        map.put("createdAt", c.getCreatedAt());
        map.put("author", Map.of("_id", c.getAuthor().getId(), "id", c.getAuthor().getId(), "name", c.getAuthor().getName()));
        return map;
    }
}
